### PR TITLE
feat: support different registration methods

### LINKED ISSUE #136: Allow to specify  service Ip to use for controlPlane
**Describe the solution you'd like:**
Right now the [getIPAddress](https://github.com/rancher-sandbox/cluster-api-provider-rke2/blob/c565c8fefe0feed1bf6774bdd73634a0fe085b23/controlplane/internal/controllers/rke2controlplane_controller.go#L737) function defined  in the rke2controlplane_controller take the first InternalIP of the machines as address to be used as the availableServerIP.

This  is is an issue when having nodes with several interfaces. Moreover with Baremetalhost the getIPAddress function get IP addresses from the machines.cluster.x-k8s.io and those addresses are the one from the introspection step of the baremetalhost.
on that case we want to  ignore addresses from the inspection for two important reasons:
-some IPs are obtained through DHCP, but when provisioning we use static IPs
-NIC names depends on the OS distribution and might differ from IPA and target OS

We have tried to use the   [advertiseAddress](https://github.com/rancher-sandbox/cluster-api-provider-rke2/blob/c565c8fefe0feed1bf6774bdd73634a0fe085b23/controlplane/config/crd/bases/controlplane.cluster.x-k8s.io_rke2controlplanes.yaml#L551) parameter from serverConfig but it has no effect

As a result we cannot control which address will be set on worker config.yaml as server and  use to join the cluster leading to wrong setup.

We should be able to specify in the serverConfig of the RKE2ControlPlane the address to be used as the AvailableServerIp and relay on an Address randomly taken on the machines

**Why do you want this feature:**
We need this feature to be able to deploy RKE2 cluster on environment with nodes with several nics

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### LINKED ISSUE #146: Failing to bootstrap control plane with v0.1.0-alpha.3
**What happened:**

While attempting to upgrade from v0.1.0-alpha.2 to v0.1.0-alpha.3 (without changing anything else) first control plane node provisionning is stuck with controlplane controller looping on following messages:

```
I0612 15:47:37.061255       1 rke2controlplane_controller.go:499]  "msg"="Initializing control plane" "Desired"=3 "Existing"=0 "RKE2ControlPlane"={"name":"management-cluster-control-plane","namespace":"default"} "controller"="rke2controlplane" "controllerGroup"="controlplane.cluster.x-k8s.io" "controllerKind"="RKE2ControlPlane" "name"="management-cluster-control-plane" "namespace"="default" "reconcileID"="7f678957-46e5-40be-9149-2d8ef42db0d7"
E0612 15:47:37.085811       1 controller.go:329]  "msg"="Reconciler error" "error"="control plane has already been initialized, found 1 owned machine for cluster default/management-cluster: controller cache or management cluster is misbehaving" "RKE2ControlPlane"={"name":"management-cluster-control-plane","namespace":"default"} "controller"="rke2controlplane" "controllerGroup"="controlplane.cluster.x-k8s.io" "controllerKind"="RKE2ControlPlane" "name"="management-cluster-control-plane" "namespace"="default" "reconcileID"="7f678957-46e5-40be-9149-2d8ef42db0d7"
```

After some investigation it turns out to be caused by the fact that cluster fails to [retrieve control plane nodes](https://github.com/rancher-sandbox/cluster-api-provider-rke2/blob/main/controlplane/internal/controllers/rke2controlplane_controller.go#LL428C10-L428C10).

Control plane nodes are filtered out, as they are labelled with `cluster.x-k8s.io/control-plane-name` instead of `cluster.x-k8s.io/control-plane`. We should use `clusterv1.MachineControlPlaneLabel` instead of `clusterv1.MachineControlPlaneNameLabel` [here](https://github.com/rancher-sandbox/cluster-api-provider-rke2/blob/main/pkg/rke2/control_plane.go#LL223C3-L223C41), as this the the label that is used to[ filter control plane machines](https://github.com/kubernetes-sigs/cluster-api/blob/main/util/collections/machine_filters.go#L224)

### PR DESCRIPTION
**Pull request by richardcase**
**What this PR does / why we need it**:

This adds support for different ways to register new nodes into the cluster.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #136
Fixes #146 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits into logical changes
- [x] includes documentation
- [x] adds unit tests

### CODE REVIEW

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go [hunk 1]
```diff
@@ -44,6 +47,12 @@ var _ webhook.Defaulter = &RKE2ControlPlane{}
 // Default implements webhook.Defaulter so a webhook will be registered for the type.
 func (r *RKE2ControlPlane) Default() {
 	bootstrapv1.DefaultRKE2ConfigSpec(&r.Spec.RKE2ConfigSpec)
+
+	if r.Spec.RegistrationMethod == RegistrationMethodAddress {
+		if r.Spec.ServerConfig.AdvertiseAddress == "" {
+			r.Spec.ServerConfig.AdvertiseAddress = r.Spec.RegistrationAddress
```
Comment (furkatgofurov7) [inline controlplane/api/v1alpha1/rke2controlplane_webhook.go:53]: can we log this properly communicate what is happening under the hood?

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go [hunk 2]
```diff
@@ -52,20 +61,48 @@ var _ webhook.Validator = &RKE2ControlPlane{}
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
 func (r *RKE2ControlPlane) ValidateCreate() error {
-	if bootstrapv1.ValidateRKE2ConfigSpec(r.Name, &r.Spec.RKE2ConfigSpec) != nil {
-		return bootstrapv1.ValidateRKE2ConfigSpec(r.Name, &r.Spec.RKE2ConfigSpec)
+	rke2controlplanelog.Info("RKE2ControlPlane validate create", "control-plane", klog.KObj(r))
```
Comment (furkatgofurov7) [inline controlplane/api/v1alpha1/rke2controlplane_webhook.go:64]:
```suggestion
	rke2controlplanelog.Info("RKE2ControlPlane validate create", "rke2-control-plane", klog.KObj(r))
```

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go [hunk 3]
```diff
@@ -52,20 +61,48 @@ var _ webhook.Validator = &RKE2ControlPlane{}
...
 
-	return ValidateRKE2ControlPlaneSpec(r.Name, &r.Spec)
+	return apierrors.NewInvalid(GroupVersion.WithKind("RKE2ControlPlane").GroupKind(), r.Name, allErrs)
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
 func (r *RKE2ControlPlane) ValidateUpdate(old runtime.Object) error {
-	if bootstrapv1.ValidateRKE2ConfigSpec(r.Name, &r.Spec.RKE2ConfigSpec) != nil {
-		return bootstrapv1.ValidateRKE2ConfigSpec(r.Name, &r.Spec.RKE2ConfigSpec)
+	rke2controlplanelog.Info("RKE2ControlPlane validate update", "control-plane", klog.KObj(r))
```
Comment (furkatgofurov7) [inline controlplane/api/v1alpha1/rke2controlplane_webhook.go:81]:
```suggestion
	rke2controlplanelog.Info("RKE2ControlPlane validate update", "rke2-control-plane", klog.KObj(r))
```

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go [hunk 4]
```diff
@@ -52,20 +61,48 @@ var _ webhook.Validator = &RKE2ControlPlane{}
...
 
-	return ValidateRKE2ControlPlaneSpec(r.Name, &r.Spec)
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, bootstrapv1.ValidateRKE2ConfigSpec(r.Name, &r.Spec.RKE2ConfigSpec)...)
+	allErrs = append(allErrs, r.validateCNI()...)
+
+	if r.Spec.RegistrationMethod != oldControlplane.Spec.RegistrationMethod {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "registrationmethod"), r.Spec.RegistrationMethod, "field is immutable"),
```
Comment (furkatgofurov7) [inline controlplane/api/v1alpha1/rke2controlplane_webhook.go:97]:
```suggestion
			field.Invalid(field.NewPath("spec", "registrationMethod"), r.Spec.RegistrationMethod, "field is immutable"),
```

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go [hunk 5]
```diff
@@ -75,24 +112,27 @@ func (r *RKE2ControlPlane) ValidateDelete() error {
 	return nil
 }
 
-// ValidateRKE2ControlPlaneSpec validates the RKE2ControlPlaneSpec Object.
-func ValidateRKE2ControlPlaneSpec(name string, spec *RKE2ControlPlaneSpec) error {
-	allErrs := spec.validate()
-	if len(allErrs) == 0 {
-		return nil
+func (r *RKE2ControlPlane) validateCNI() field.ErrorList {
```
Comment (furkatgofurov7) [inline controlplane/api/v1alpha1/rke2controlplane_webhook.go:115]: godoc is missing

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go [hunk 6]
```diff
@@ -75,24 +112,27 @@ func (r *RKE2ControlPlane) ValidateDelete() error {
...
+				r.Spec.ServerConfig.CNI, "must be specified when cniMultusEnable is true"))
 	}
 
-	return apierrors.NewInvalid(GroupVersion.WithKind("RKE2ControlPlane").GroupKind(), name, allErrs)
+	return allErrs
 }
 
-// validate validates the RKE2ControlPlaneSpec Object.
-func (s *RKE2ControlPlaneSpec) validate() field.ErrorList {
+func (r *RKE2ControlPlane) validateRegistrationMethod() field.ErrorList {
```
Comment (furkatgofurov7) [inline controlplane/api/v1alpha1/rke2controlplane_webhook.go:127]: godoc is missing

File: controlplane/api/v1alpha1/types.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+// RegistrationMethod defines the methods ro use for registering a new node in a cluster.
```
Comment (furkatgofurov7) [inline controlplane/api/v1alpha1/types.go:19]:
```suggestion
// RegistrationMethod defines the methods to use for registering a new node in a cluster.
```

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 8]
```diff
@@ -352,15 +353,14 @@ func (r *RKE2ControlPlaneReconciler) updateStatus(ctx context.Context, rcp *cont
...
-	validIPAddresses := []string{}
-
-	for _, machine := range availableCPMachines {
-		ipAddress, err := getIPAddress(*machine)
-		if err != nil {
-			break
-		}
+	registrationmethod, err := registration.NewRegistrationMethod(string(rcp.Spec.RegistrationMethod))
+	if err != nil {
+		return fmt.Errorf("getting node registration method: %w", err)
```
Comment (furkatgofurov7) [inline controlplane/internal/controllers/rke2controlplane_controller.go:358]:
```suggestion
		return fmt.Errorf("error getting node registration method: %w", err)
```

File: controlplane/internal/controllers/rke2controlplane_controller.go [hunk 9]
```diff
@@ -352,15 +353,14 @@ func (r *RKE2ControlPlaneReconciler) updateStatus(ctx context.Context, rcp *cont
...
+	}
 
-		validIPAddresses = append(validIPAddresses, ipAddress)
+	validIPAddresses, err := registrationmethod(rcp, availableCPMachines)
+	if err != nil {
+		return fmt.Errorf("getting registration addresses: %w", err)
```
Comment (furkatgofurov7) [inline controlplane/internal/controllers/rke2controlplane_controller.go:363]:
```suggestion
		return fmt.Errorf("error getting registration addresses: %w", err)
```

**Review by furkatgofurov7 — APPROVED** [review]
Great work, few non blocking nits:

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go (hunk 2 above)
Comment (alexander-demicev) [reply controlplane/api/v1alpha1/rke2controlplane_webhook.go:64]: should we remove this log line? It's only telling that webhook is validating a resource

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go (hunk 1 above)
Comment (richardcase) [reply controlplane/api/v1alpha1/rke2controlplane_webhook.go:53]: Added a log line.

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go (hunk 4 above)
Comment (richardcase) [reply controlplane/api/v1alpha1/rke2controlplane_webhook.go:97]: Changed.

File: controlplane/api/v1alpha1/rke2controlplane_webhook.go (hunk 3 above)
Comment (richardcase) [reply controlplane/api/v1alpha1/rke2controlplane_webhook.go:81]: Removed this log line

File: controlplane/api/v1alpha1/types.go (hunk 7 above)
Comment (richardcase) [reply controlplane/api/v1alpha1/types.go:19]: Changed, thanks.

File: controlplane/internal/controllers/rke2controlplane_controller.go (hunk 8 above)
Comment (richardcase) [reply controlplane/internal/controllers/rke2controlplane_controller.go:358]: If i'm wrapping error i generally don'y include "error" but instead give the context.

**Review by salasberryfin — APPROVED** [review]
LGTM

**Review by furkatgofurov7 — APPROVED** [review]

**Pull request merged**